systick.cpp

```diff
@@ -110,23 +156,126 @@ unsigned long t_systick1 = 0;
+ * should ideally last no more than 10us.
+ * 
+ * The reason fo rthis apparently arcane technique is that the 
+ * ATMEGA328P does not have nested interrupts. At top speed, the 
+ * encoders may generate interrupt at more than 20kHz and we cannot
+ * afford to miss one. Also, received serial characters must be serviced
+ * as they arrive or they will be lost. Having a single systick event
+ * that does everything but takes many tens of microseconds risks
+ * lost serial data and encoder pulses.
+ *  
+ * For testing, the built-in LED is turned on at the start of the interrupt
+ * and off again at the end. This can be used to measure system load.
+ * Disable the feature if you want to use the built-in LED for some other 
+ * purpose.
+ */ 
+ISR(TIMER2_COMPA_vect) {
+  digitalWriteFast(LED_BUILTIN, 1);
+  systick_phase++;
+  if (systick_phase >= 40) {
+    systick_phase = 0;
+  }
+  switch (systick_phase) {
+    case 0:
+      // always start conversions as soon as  possible so they get a 
+      // full 50us to convert
+      start_adc(BATTERY_VOLTS);
+      delayMicroseconds(40);  // just a long marker for the oscilloscope.
+      break;
+    case 1:
+      raw_BatteryVolts_adcValue = get_adc_result();
+      start_adc(FUNCTION_PIN);
+      // TODO find a magic voltage divider ratio that makes this a single multiply for millivolts
+      raw_BatteryVolts = (raw_BatteryVolts_adcValue * 2 * 5) / 1024;
+      break;
+    case 2:
+      Switch_ADC_value = get_adc_result();
+      start_adc(A0);
+      break;
+    case 3:
+      gSensorA0_dark = get_adc_result();
+      start_adc(A1);
+      break;
+    case 4:
+      gSensorA1_dark = get_adc_result();
+      start_adc(A2);
+      break;
+    case 5:
+      gSensorA2_dark = get_adc_result();
+      start_adc(A3);
+      break;
+    case 6:
+      gSensorA3_dark = get_adc_result();
+      start_adc(A4);
+      break;
+    case 7:
+      gSensorA4_dark = get_adc_result();
+      start_adc(A5);
+      break;
+    case 8:
+      gSensorA5_dark = get_adc_result();
+      // got all the dark ones so light them up
+      if (emitter_on) {
+        digitalWriteFast(EMITTER, 1);
+      }
+      // wait at least one cycle for the detectors to respond
+      break;
+    case 9:
+      start_adc(A0);
+      break;
+    case 10:
+      gSensorA0_light = get_adc_result();
+      start_adc(A1);
+      break;
+    case 11:
+      gSensorA1_light = get_adc_result();
+      start_adc(A2);
+      break;
+    case 12:
+      gSensorA2_light = get_adc_result();
+      start_adc(A3);
+      break;
+    case 13:
+      gSensorA3_light = get_adc_result();
+      start_adc(A4);
+      break;
+    case 14:
+      gSensorA4_light = get_adc_result();
+      start_adc(A5);
+      break;
+    case 15:
+      gSensorA5_light = get_adc_result();
+      if (emitter_on) {
+        digitalWriteFast(EMITTER, 0);
+      }
+      break;
+    default:
+      break;
+  }
+  digitalWriteFast(LED_BUILTIN, 0);
```

Comment:
Testing line?

Reply:
As for previous LED flashing - it can go.